1382 browse tab wrong focus on click

juzen2003 opened the pull request:
- Fixes #1382
- Is a new database import required? N
- Were test and deploy scripts reviewed for necessary changes? NA
- Were any JavaScript or CSS files modified? Y
  - If YES:
    - JSHINT run on all affected files: Y
    - Tested on Chrome, Firefox, Safari, Edge, and iOS: Y
      (Remember to test all browser sizes)
    - Tested for race conditions (with delayed API calls if applicable): NA
- Was the documentation reviewed for necessary changes or additions? NA
  - About
  - Getting Started
  - FAQ
  - API Guide
  - Tooltips
  - Import/database schema

Description of changes:

- For perfect scrollbar to handle any keydown event, we have to make sure the ps container is hovered and perfect scrollbar is focused. This is the required step from the source code (`perfect-scrollbar.js` line  629-631). So whenever we click or hit up/down arrows, and page up/down keys in browse tab, we make sure the ps container is hovered and ps is focused so the the keydown event will be properly handled by ps, and the observations view can be properly scrolled.
- For record, if we undo all the changes in `browse.js`, and just comment out line 629-631 in `perfect-scrollbar.js`, we can achieve the same behavior.
- Update on 7/26/24: 
  - Disable the up/down and left/right arrow keys to switch between different tabs when we click on the top nav bar.
  - Now in the browse view, the up/down and page up/down keys will only control the perfect scrollbar of the observations view, and it will work no matter where the mouse cursor is on the screen.
  - Test on Windows 11 with Chrome (119-126), Firefox (120-128), and Edge (119-126). Mac Sonoma latest Chrome, Firefox, and Safari.

Known problems:
 
 <div id='description'>
<h3>Summary by Bito</h3>
The changes primarily focus on improving user interface interactions and scroll behavior. In the CSS file, styles were added to remove highlight borders on focus. The JavaScript files saw multiple updates to handle scroll positions, event bindings, and synchronization of scrollbar positions. Additionally, navigation click handling and focus management were enhanced to improve user experience.
<br>
<br>
<b>Code change type</b>: Refactoring, Performance Improvement, Accessibility, Functionality
<br>
<br>
<b>Unit tests added</b>: False
<br>
<br>
<b>Estimated effort to review (1-5, lower is better)</b>: 2
</div>

juzen2003 committed:
In browse tab, whenever we click or hit up/down, and page up/down keys,
make sure the ps container is hovered and ps is focused so the the
keydown event will be properly handled by ps.
- In browse.js (line 664-677), create hoverAndFocusOnPS to mouse over
  the ps container and focus on ps, this is the required action for ps
  to handle the keydown event.
- In browse.js (line 578-579), make sure slider won't scroll when the
  up/down arrow keys are hit.
- In opus.css (273-276), remove the highlighted border of perfect
  scrollbar when it's focused.

juzen2003 committed:
Add missing semicolon at line 579,588,675 and 676 in browse.js

juzen2003 committed:
When clicking nav bar to switch between tabs, we focus out the tab so
that up/down and left/right arrow keys won't switch the tabs.

juzen2003 committed:
Remove the commented out code in opus.js.

juzen2003 committed:
Remove the commented out ..hoverAndFocusOnPS function calls in browse.js

juzen2003 committed:
Make sure browse gallery and table view has slider number in sync
- In browse.js (line 139-144), add a step to make sure trFloor is
  properly set for calculations.
- In browse.js (line 1749-1754, 1729-1734), sync up scrollbar position
  when switching between gallery & table view.

juzen2003 committed:
Sync up the behavior of slider and scrollbar position when switching
between the browse gallery and table views.
- In browse.js (line 1738-1746), manually set the slider back to the
  first obs of the first row in the gallery view when we switch from
  the table view to the gallery view
- In browse.js (line 1765-1767), when switching from gallery to table
  view, make sure scrollbar postion at the table view is always pointing
  at the first obs of the first row in gallery view before switching.

juzen2003 committed:
Removed the commented out code in browse.js

juzen2003 committed:
Make sure when switching from the gallery view to the table view,
scrollbar position in the table view is always at the first obs of the
first row in the gallery view before switching
- In brwose.js (line 1763-1769), add the step to calculate the first
  obs of the first row in the gallery view before setting the scrollbar
  position in the table view.

rfrenchseti reviewed (changes requested):
Two issues:
1) When you open any Help dialog (like About or Bundles/Volumes), using the up/down or page up/page down keys still scroll the thumbnails in the background, not the text in the dialog.
2) When you switch from Gallery to Table view, the top line of the Table view is not the same image number as the top-left thumbnail in the Gallery (in other words, the slider moves when you switch views).

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -136,8 +136,11 @@ var o_browse = {
-            window.scrollTo(0, 0); // restore previous scroll position
+            // Make sure trFloor is properly set for calculations
+            let viewNamespace = opus.getViewNamespace()
+            if (viewNamespace.galleryBoundingRect.trFloor===Infinity) {
```
We normally have spaces around operators like `===`

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -575,7 +578,19 @@ var o_browse = {
+                (e.which || e.keyCode) == 33 || (e.which || e.keyCode) == 34)) {
+                o_browse.hoverAndFocusOnPS();
+            }
+
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/SETI/rms-opus/pull/1387#issuecomment-2276954910>#c06976</a><span> - 08/09/2024, 12:49 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The use of 'e.which' and 'e.keyCode' is deprecated and may not be supported in future versions of browsers. This can lead to potential issues with cross-browser compatibility and future-proofing the code.
</div>
<div id="fix">
<b>Fix:</b> Replace the deprecated 'e.which' and 'e.keyCode' with 'e.key' or 'e.code' to ensure future compatibility and adherence to modern standards.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -588,6 +588,6 @@
              if (opus.prefs.view === "browse" && (e.type === "click" ||
 -                (e.which || e.keyCode) == 38 || (e.which || e.keyCode) == 40 ||
 -                (e.which || e.keyCode) == 33 || (e.which || e.keyCode) == 34)) {
 +                e.key === "ArrowUp" || e.key === "ArrowDown" ||
 +                e.key === "PageUp" || e.key === "PageDown")) {
                  o_browse.hoverAndFocusOnPS();
              }
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -649,6 +664,21 @@ var o_browse = {
         });
     }, // end browse behaviors
 
+    // Hover to ps container and focus on ps, this is the key step to make sure ps listen
+    // to the keydown (up/down arrows, page up, page down, and etc) event. By default, if
+    // the cursor moves out of the ps container or the ps is not focused, ps won't handle
+    // any keydown event (perfect-scrollbar.js line 629-631). Note: Without this function,
+    // we can just comment out the code at line 629-631 in perfect-scrollbar.js to get the
+    // same behavior.
+    hoverAndFocusOnPS: function() {
+        let view = opus.prefs.view;
+        let tab = opus.getViewTab(view);
+        let contentsView = o_browse.getScrollContainerClass(view);
+
+        $(`${tab} ${contentsView}`).mouseover();
+        $(`${tab} ${contentsView} .ps__thumb-y`).focus();
+    },
+
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/SETI/rms-opus/pull/1387#issuecomment-2276954910>#c06976</a><span> - 08/09/2024, 12:49 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The use of jQuery's '.mouseover()' and '.focus()' methods can lead to unexpected behavior, especially in dynamic content scenarios. This can cause performance issues and potential memory leaks if not handled properly.
</div>
<div id="fix">
<b>Fix:</b> Consider using native JavaScript methods to handle the focus and mouseover events. Additionally, ensure that these events are properly cleaned up to avoid memory leaks.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -678,7 +678,7 @@
         let view = opus.prefs.view;
         let tab = opus.getViewTab(view);
         let contentsView = o_browse.getScrollContainerClass(view);

 -       $('${tab} ${contentsView}').mouseover();
 -       $('${tab} ${contentsView} .ps__thumb-y').focus();
 +       const container = document.querySelector('${tab} ${contentsView}');
 +       const thumb = container.querySelector('.ps__thumb-y');
 +       if (container && thumb) {
 +           container.dispatchEvent(new Event('mouseover'));
 +           thumb.focus();
 +       }
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1048,8 +1078,8 @@ var o_browse = {
         // than rest of tr, and this will mess up the calculation.
-        let tableRowHeight = ($(`${tab} tbody tr`).length === 1 ? $(`${tab} tbody tr`).outerHeight() :
-                              $(`${tab} tbody tr`).eq(1).outerHeight());
+        let tableRowHeight = ($(`${tab} tbody tr`).length === 1 ? $(`${tab} tbody tr td`).outerHeight() :
+                              $(`${tab} tbody tr td`).eq(1).outerHeight());
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/SETI/rms-opus/pull/1387#issuecomment-2276954910>#c06976</a><span> - 08/09/2024, 12:49 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The change from 'tbody tr' to 'tbody tr td' in the jQuery selector may lead to incorrect height calculations. The original code calculates the height of the table row, while the updated code calculates the height of the table cell. This could lead to layout issues if the table cells have different heights or if there are multiple cells in a row.
</div>
<div id="fix">
<b>Fix:</b> Revert the change to ensure the height of the table row is calculated correctly. If the intention is to handle single-cell rows differently, additional logic should be added to handle that case specifically.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -1081,2 +1081,2 @@
 - let tableRowHeight = ($('${tab} tbody tr').length === 1 ? $('${tab} tbody tr td').outerHeight() :
 - $('${tab} tbody tr td').eq(1).outerHeight());
 + let tableRowHeight = ($('${tab} tbody tr').length === 1 ? $('${tab} tbody tr').outerHeight() :
 + $('${tab} tbody tr').eq(1).outerHeight());
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

rfrenchseti commented on opus/application/static_media/css/opus.css:
```diff
@@ -270,6 +270,11 @@ so that both "Recent Announcements" and "Message" can be vertically aligned. */
 }
 
+/* remove the highlight border of ps when it's focus */
+.op-gallery-contents .ps__thumb-y:focus {
+    outline: none;
+}
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/SETI/rms-opus/pull/1387#issuecomment-2276954910>#c06976</a><span> - 08/09/2024, 12:49 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> Removing the outline on focus can negatively impact accessibility. Users who rely on keyboard navigation may find it difficult to determine which element is currently focused.
</div>
<div id="fix">
<b>Fix:</b> Instead of removing the outline entirely, consider using a custom outline style that maintains accessibility while achieving the desired visual effect.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -274,3 +274,3 @@
  .op-gallery-contents .ps__thumb-y:focus {
 -    outline: none;
 +    outline: 2px solid #000; /* Example: Use a custom outline style */
  }
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

rfrenchseti commented on opus/application/static_media/js/opus.js:
```diff
@@ -802,6 +802,9 @@ var opus = {
 
+            // Focus out of the nav bar so that up/down and left/right arrow keys won't
+            // jump between different tabs
+            e.target.blur();
```
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/SETI/rms-opus/pull/1387#issuecomment-2276954910>#c06976</a><span> - 08/09/2024, 12:49 am</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The use of 'e.target.blur()' to remove focus from the nav bar might cause unexpected behavior if the focus is needed elsewhere. This can lead to accessibility issues, especially for keyboard-only users who rely on focus indicators to navigate.
</div>
<div id="fix">
<b>Fix:</b> Instead of blurring the target element, consider managing focus more explicitly. For example, set focus to a specific element that should receive it after the nav bar loses focus. This ensures that the user experience remains consistent and accessible.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -804,0 +805,5 @@ var opus = {
 +            // Focus out of the nav bar so that up/down and left/right arrow keys won't
 +            // jump between different tabs
 +            // Instead of blurring, set focus to a specific element
 +            document.getElementById('desired-element-id').focus();
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

juzen2003 committed:
Make sure up/down & page up/down keys can scroll the help panel when the
help panel is open in the browse view.
- In opus.js (line 825-830), make sure keydown event triggered from help
  panel won't propagate to galler/table views. This will make sure
  up/down & page up/down keys scroll the help panels instead of gallery
  and table views.
- In opus.js (line 1142-1145), make sure to focus the ps scrollbar in
  help panel so that it will handle the keydown event.

juzen2003 committed:
Make sure slider number matches to the first obs of the first row in
the gallery view when switching between the gallery/table views.

juzen2003 committed:
Fixed behavior of slider and ps scrollbar
- In browse.js (line 1296-1298), if startObs is undefined, we can just
  return, this will avoid showing NaN on slider (instead just shows
  spinner).
- In browse.js (line 1803-1807), update ps scrollbar so that it will
  show up at the first gallery/table views switching after refreshing
  the page.
- In browse.js (line 1275-1285), make sure we don't trigger ps-scroll-up
  or set scrollbar position when we are at the top obs. This will avoid
  some weird twitching when trying to scroll up to obs 1.

juzen2003 committed:
Clean up js warning flagged by jshint.

juzen2003 committed:
Replace e.which and e.keycode to e.code in browse.js & opus.js

juzen2003 committed:
Make sure keyboard can control the help panel ps when the help panel is
open in the browse tab.
- In browse.js (line 604-610), when the ps in help panel shows up, we
we don't hover & focus to the gallery/table views ps if there is a help
panel ps on the screen.

rfrenchseti reviewed (changes requested):
The problem with the arrow keys being used while on the Browse tab with a Help pane open is only partially fixed. Now when go to the Browse tab, click on a thumbnail to get a slideshow, and then open the Help, the Help pane won't scroll using the keys. But if there's no slideshow open, the Help pane works properly.

rfrenchseti reviewed (changes requested):
On the Cart tab (with a lot of items in the cart), if you drag the slider and leave the mouse over the slider, the up/down arrow keys don't do anything. But they work if you move the mouse into the thumbnail browse area.

juzen2003 committed:
Make sure when the help panel is open while slide show is also open,
arrow keys are controlling the help panel.

juzen2003 committed:
Merge remote-tracking branch 'upstream/main' into 1382_browse_tab_wrong_focus_on_click

juzen2003 committed:
Fixed the code error (typo) at line 625 in browse.js

juzen2003 committed:
In cart tab, make sure when slider is focused, arrow keys can properly
scroll the gallery view.

rfrenchseti reviewed (approved):
There are still some issues with the Cart tab that should be addressed separately:

- Drag the slider and release; press up or down arrow once and nothing happens; press a second time and the scrollbar moves
- Drag the slider and while dragging move the mouse far off vertically (like to the lower-right corner of the browser); when you release, the slider will snap back to where it started from. This doesn't happen on the Browse tab.
- Make the browser small so the Download Data pains becomes a pop-out. Pop it out. Now the arrow keys no longer doing anything on that pane (scrolling the list of products), they always move the thumbnails in the background.
